internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,671 @@
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "PostgresSecretFailure",
+			Message: "Failed to generate database secret: " + err.Error(),
+		})
+		return err
+	}
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, secret, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "PostgresSecretFailure",
+			Message: "Failed to ensure database secret: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Database secret created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureAssistedServiceDeployment(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	// must have the route in order to populate SERVICE_BASE_URL for the service
+	route := &routev1.Route{}
+	err := r.Get(ctx, types.NamespacedName{Name: name, Namespace: r.Namespace}, route)
+	if err != nil || route.Spec.Host == "" {
+		if err == nil {
+			err = fmt.Errorf("Route's host is empty")
+		}
+		r.Log.Info("Failed to get route or route's host is empty")
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "DeploymentFailure",
+			Message: "Failed to get route for assisted service: " + err.Error(),
+		})
+		return err
+	}
+
+	deployment, mutateFn := r.newAssistedServiceDeployment(instance, route.Spec.Host)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, deployment, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "DeploymentFailure",
+			Message: "Failed to ensure assisted service deployment: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Assisted service deployment created")
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *AgentServiceConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&adiiov1alpha1.AgentServiceConfig{}).
+		Owns(&corev1.PersistentVolumeClaim{}).
+		Owns(&corev1.Service{}).
+		Owns(&corev1.Secret{}).
+		Owns(&routev1.Route{}).
+		Owns(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func (r *AgentServiceConfigReconciler) newFilesystemPVC(instance *adiiov1alpha1.AgentServiceConfig) (*corev1.PersistentVolumeClaim, controllerutil.MutateFn) {
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-filesystem", agentServiceConfigName),
+			Namespace: r.Namespace,
+		},
+		Spec: instance.Spec.FileSystemStorage,
```

Comment:
I think it would be wise to make a copy of this so as not to have unexpected side effects in the cache.

Reply:
Oops it's not a pointer, so never mind on this one.